src/TypeShape.Examples/MsgPackSerializer/MsgPackSerializer.cs

```diff
@@ -0,0 +1,345 @@
+                T result = constructor();
+
+                for (int i = 0; i < count; i++)
+                {
+                    ReadOnlySpan<byte> propertyName = MessagePack.Internal.CodeGenHelpers.ReadStringSpan(ref reader);
+                    if (propertyReaders.TryGetValue(propertyName, out DeserializeProperty<T>? propertyReader))
+                    {
+                        propertyReader(ref result, ref reader, options);
+                    }
+                    else
+                    {
+                        reader.Skip();
+                    }
+                }
+
+                return result;
+            };
+
+            FormatSerialize<T> serialize = (ref MessagePackWriter writer, T? value, MessagePackSerializerOptions options) =>
+            {
+                if (value is null)
+                {
+                    writer.WriteNil();
+                    return;
+                }
+
+                writer.WriteMapHeader(properties.Count);
+                foreach (var property in properties)
+                {
+                    writer.WriteRaw(property.RawPropertyNameString);
+                    property.Serialize(ref value, ref writer, options);
+                }
+            };
+
+            return new Formatter<T>(deserialize, serialize);
+        }
+
+        public override object? VisitProperty<TDeclaringType, TPropertyType>(IPropertyShape<TDeclaringType, TPropertyType> propertyShape, object? state = null)
+        {
+            var formatter = GetFormatter(propertyShape.PropertyType);
+
+            Getter<TDeclaringType, TPropertyType> getter = propertyShape.GetGetter();
+            SerializeProperty<TDeclaringType> serialize = (ref TDeclaringType container, ref MessagePackWriter writer, MessagePackSerializerOptions options) =>
+            {
+                formatter.Serialize(ref writer, getter(ref container), options);
+            };
+
+            Setter<TDeclaringType, TPropertyType> setter = propertyShape.GetSetter();
+            DeserializeProperty<TDeclaringType> deserialize = (ref TDeclaringType container, ref MessagePackReader reader, MessagePackSerializerOptions options) =>
+            {
+                setter(ref container, formatter.Deserialize(ref reader, options)!);
+            };
+
+            return (serialize, deserialize);
+        }
+
+        public override object? VisitConstructor<TDeclaringType, TArgumentState>(IConstructorShape<TDeclaringType, TArgumentState> constructorShape, object? state = null)
+        {
+            return constructorShape.GetDefaultConstructor();
+        }
+    }
+
+    private delegate T? FormatDeserialize<T>(ref MessagePackReader reader, MessagePackSerializerOptions options);
+    private delegate void FormatSerialize<T>(ref MessagePackWriter writer, T? value, MessagePackSerializerOptions options);
+
+    private sealed class Formatter<T>(FormatDeserialize<T> deserialize, FormatSerialize<T> serialize) : IMessagePackFormatter<T?>
```

Comment:
I can see why you went for a delegate adapter, but it does result in two layers of indirection for every call, which adds up as the object graph is being traversed.

Reply:
Fixed.
Can you review all the places where `formatters.GetOrAdd` appears in my code and tell me if I'm doing it right? I don't understand why this pattern is required, so I'm not sure...

Reply:
I was suggesting that you could make `Formatter<T>` an abstract class and have individual types and shapes inherit from that. The downside is that you need to factor those into individual classes as opposed to inlining lambda expressions, but on the other hand splitting your code into a folder of Formatters and then a keeping a separate builder that folds them together should make things more manageable.

Reply:
So far, I'm not sure what I'd do with more than the two implementations I have.
Maybe I'll need more as I get to supporting non-default constructors, etc. though.